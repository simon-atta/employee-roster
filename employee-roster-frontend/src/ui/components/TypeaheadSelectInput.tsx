/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import Select, { CommonProps, components } from 'react-select';

export interface TypeaheadSelectProps<T> {
  emptyText: string;
  options: T[];
  value: T | undefined;
  optionToStringMap: (option: T) => string;
  onChange: (selected: T | undefined) => void;
  optional?: boolean;
  noClearButton?: boolean;
  autoSize?: boolean;
  valueComponent?: React.FC<CommonProps<{ value: T}> & { data: {value: T}}>;
}

export default class TypeaheadSelectInput<T> extends React.Component<TypeaheadSelectProps<T>> {
  constructor(props: TypeaheadSelectProps<T>) {
    super(props);
    this.onSelect = this.onSelect.bind(this);
  }

  onSelect(selection: { value: T }|undefined) {
    this.props.onChange(selection ? selection.value : undefined);
  }

  render() {
    const { optionToStringMap, emptyText, optional, options, autoSize } = this.props;
    const selectOptions = this.props.options.map(o => ({ value: o }));
    const Display = this.props.valueComponent;
    const SingleValue = Display || components.SingleValue;
    const Option = Display ? (props: any, children: any) => (
      <components.Option {...props}>
        <Display {...props}>{children}</Display>
      </components.Option>
    )
      : components.Option;

    // Why map to string first?
    // Clone of objects are not the same, and sometimes we are passed
    // clones instead of the real thing, and "===" and "==" are
    // both identity/is comparators on objects, so we compare their
    // "toString" values to check for equality
    const selected = (this.props.value !== undefined) ? selectOptions
      .find(o => optionToStringMap(o.value) === optionToStringMap(this.props.value as T)) : undefined;
    const parentStyle: React.CSSProperties = {
      position: 'relative',
    };
    if (autoSize === undefined || autoSize) {
      const maxLabelLength = Math.max(emptyText.length, ...options.map(optionToStringMap).map(s => s.length));
      parentStyle.width = `${maxLabelLength + 2}em`;
    }

    return (
      <div style={parentStyle}>
        <Select
          aria-label={emptyText}
          styles={{
            menuPortal: provided => ({
              ...provided,
              zIndex: 9999999,
            }),
            singleValue: provided => ({
              ...provided,
            }),
          }}
          components={{ SingleValue, Option }}
          onChange={this.onSelect}
          defaultValue={selected}
          value={selected}
          selected={selected}
          placeholder={emptyText}
          getOptionLabel={o => optionToStringMap(o.value)}
          getOptionValue={o => optionToStringMap(o.value)}
          options={selectOptions}
          menuPortalTarget={document.body}
          isClearable={optional}
        />
      </div>
    );
  }
}
