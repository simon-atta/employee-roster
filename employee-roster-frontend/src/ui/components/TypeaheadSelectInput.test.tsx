/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import TypeaheadSelectInput, { TypeaheadSelectProps } from './TypeaheadSelectInput';

interface MockData {
  name: string;
}

type Select = TypeaheadSelectInput<MockData>;
describe('TypeaheadSelectInput component', () => {
  it('should set selected and call onChange', () => {
    jest.useFakeTimers();
    const defaultValue = { name: 'Option 2' };
    const select = mount(<TypeaheadSelectInput {...selectProps} value={defaultValue} />);
    select.setState({ isExpanded: true });
    (select.instance() as Select).onSelect({ value: { name: 'Option 1' } });
    expect(selectProps.onChange).toBeCalled();
    expect(selectProps.onChange).toBeCalledWith({ name: 'Option 1' });
  });

  it('should set selected to undefined on clear selection and call onChange', () => {
    const defaultValue = { name: 'Option 2' };
    const select = mount(<TypeaheadSelectInput {...selectProps} value={defaultValue} />);
    select.setState({ isExpanded: true });
    (select.instance() as Select).onSelect(undefined);
    expect(selectProps.onChange).toBeCalled();
    expect(selectProps.onChange).toBeCalledWith(undefined);
  });

  it('should render correctly', () => {
    const select = shallow(<TypeaheadSelectInput {...selectProps} />);
    expect(toJson(select)).toMatchSnapshot();
  });
});

const selectProps: TypeaheadSelectProps<MockData> = {
  emptyText: 'Enter some data',
  options: [{ name: 'Option 1' }, { name: 'Option 2' }, { name: 'Option 3' }],
  value: undefined,
  optionToStringMap: jest.fn(option => option.name),
  onChange: jest.fn(),
};
