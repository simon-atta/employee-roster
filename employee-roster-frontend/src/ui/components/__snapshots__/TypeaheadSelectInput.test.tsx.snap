// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeaheadSelectInput component should render correctly 1`] = `
<div
  style={
    Object {
      "position": "relative",
      "width": "17em",
    }
  }
>
  <StateManager
    aria-label="Enter some data"
    components={
      Object {
        "Option": [Function],
        "SingleValue": [Function],
      }
    }
    defaultInputValue=""
    defaultMenuIsOpen={false}
    defaultValue={null}
    getOptionLabel={[Function]}
    getOptionValue={[Function]}
    menuPortalTarget={<body />}
    onChange={[Function]}
    options={
      Array [
        Object {
          "value": Object {
            "name": "Option 1",
          },
        },
        Object {
          "value": Object {
            "name": "Option 2",
          },
        },
        Object {
          "value": Object {
            "name": "Option 3",
          },
        },
      ]
    }
    placeholder="Enter some data"
    styles={
      Object {
        "menuPortal": [Function],
        "singleValue": [Function],
      }
    }
  />
</div>
`;
